modules/MvwmConsole/MvwmConsole.1.in
.\" t
.\" @(#)@PACKAGE@-@VERSION@ @RELDATELONG@
.TH MvwmConsole 1 "@RELDATELONG@ (@VERSION@)" Mvwm "Mvwm Modules"
.UC

.SH NAME
MvwmConsole \- an mvwm command input interface

.SH SYNOPSIS
Module MvwmConsole [options]

\fIMvwmConsole\fP can only be invoked by mvwm.
Command line invocation of the \fIMvwmConsole\fP module will not work.

.SH DESCRIPTION
MvwmConsole allows the user to type mvwm configuration commands
interactively, and have them executed immediately.  This tool is
particularly useful for testing new configuration ideas, or for
implementing temporary changes to your environment.

.SH INVOCATION
MvwmConsole must be spawned as a module by mvwm.  MvwmConsole takes all
\fIxterm\fP(1) options.
.PP
MvwmConsole can be invoked by inserting the line 'Module MvwmConsole' in
the .mvwm2rc file.  This can be placed on a line by itself, if
MvwmConsole is to be spawned during mvwm's initialization, or can be
bound to a menu or mouse button or keystroke to invoke it later.

.SH CONFIGURATION OPTIONS
MvwmConsole uses \fIxterm\fP(1).  All resources set for xterm are
inherited unless overridden by command line options.

.RS
Module MvwmConsole -g 40x10 -fg black -bg green3
.RE

A different terminal emulator can be specified with the -terminal
option.  However, only terminal programs that understand the options
-name, -title and -e can be used.

.RS
Module MvwmConsole -terminal rxvt
.RE

Previous versions of MvwmConsole supported a -e option to choose a
different front-end.  Although this option is still provided for
backwards compatibility its use is discouraged unless you know
exactly what you are doing.

.RS
Module MvwmConsole -e MvwmConsoleC.pl
.RE

(see \fIMvwmConsoleC.pl\fP(1)).
.PP
Also X resources can be set in your ~/.Xdefaults file:

.RS
MvwmConsole*VT100*geometry: 40x4
.br
MvwmConsole*font: 7x14
.RE


.SH "COMMAND EDITING"

There are a few options.  If the GNU readline library is available, it
can be used.
.PP
If Perl5 is installed, MvwmConsoleC.pl can be used as a command editor.
This can be accomplished by either copying MvwmConsoleC.pl
to mvwmlib directory as MvwmConsoleC or
invoking MvwmConsole with -e option.
For example:

.RS
Module MvwmConsole -e MvwmConsoleC.pl
.RE

If neither one is installed, a simple input reading function which
doesn't have editing capabilities is used.
.P
GNU readline and MvwmConsoleC.pl have some frequent used commands
in common as default.
These commands are similar to emacs.
For more details, refer GNU readline man and info pages, and
MvwmConsoleC.pl man page.

.RS
.PD 0
.TP 8
Ctrl-A
- beginning of line
.TP
Ctrl-B
- previous char
.TP
Ctrl-D
- delete char
.TP
Ctrl-E
- end of line
.TP
Ctrl-F
- next char
.TP
Ctrl-H
- backspace
.TP
Ctrl-K
- erase to the end of line
.TP
Ctrl-N
- next line
.TP
Ctrl-P
- previous line
.TP
Ctrl-R
- search reverse
.TP
Ctrl-U
- delete line
.TP
Meta-B
- previous word
.TP
Meta-F
- next word
.TP
Esc <
- beginning of history
.TP
Esc >
- end of history
.RE
.PD

.SH EXITING

MvwmConsole can be stopped by entering the EOF character (usually CTRL-D).
.PP
\fBNote!\fP Do not use the "quit" command, as this is an mvwm builtin:
typing "quit" at the MvwmConsole command line will cause mvwm to exit.

.SH SEE ALSO
\fIxterm\fP(1), \fIMvwmConsoleC.pl\fP(1), GNU Readline library

.SH AUTHOR
MvwmConsole is the original work of Toshi Isogai.


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in release 2.6.6 (not released yet)

* Released features:

  - FvwmForm supports separator lines.
  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.

  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-desktop-menu is re-written and provides better support of
    the XDG menu specification.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take foxus themselves (stlye
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.

  - FvwmForm now honours a Xinerama screen specifier in its
    geometry.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Several possible format string attacks have been fixed.

-------------------------------------------------------------------
Changes in release 2.6.5 (20-Apr-2012)

* New features:

  - FvwmPager now wraps window names if SmallFont is set.

* Bug fixes:

  - Made signals registered with FVWM unblocking.
  - The "UnderMouse" option to PositionPlacement now honors the EWMH
    working area by default, if it's in use.
  - FvwmButtons handles the deleted window correctly for sub-windows.

-------------------------------------------------------------------
Changes in stable release 2.6.4 (01-Feb-2012)

* New features:

  - FvwmIconMan no longer allows for itself to be transient by using:

    Module FvwmIconMan Transient

    This is too ambiguous with valid module alias names, hence a
    transient FvwmIconMan must be created with "-Transient" as in:

    Module FvwmIconMan -Transient
  - New command InfoStore -- to store key/value pairs of information
    to relieve the burden of the SetEnv command.
  - Speed-up improvements for fvwm-menu-desktop when generating XDG
    menus by removing unnecessary stat(2) calls.
  - BusyCursor and CursorStyle have been set to make the cursor look
    like a dot during Read/PipeRead commands, by default.

* Bug fixes:

  - FvwmRearrange now understands the "ewmhiwa" option when placing
    windows.
  - Client gravity for subwindows is honoured for reparenting
    windows, such as with XEmbed, when changing the parent's
    geometry.
  - Conditional command processing of !Layer n, as in:

        Next (Xteddy, !Layer 4) Echo cuddles

    has now been fixed.
  - Fixed handling of swallowed windows for transient FvwmButtons'.
  - fvwm-menu-desktop now looks in /usr/share/applications for KDE
    legacy mode.
  - fvwm-config no longer accepts "--is-stable", "--is-released",
    "--is-final".

-------------------------------------------------------------------
Changes in stable release 2.6.3 (30-Sep-2011)

* New features:

  - New Style commands:

        TitleFormat
        IconTitleFormat

        Used to specify a window's visible name.  These new options
        deprecate the older styles:

        IndexedWindowName
        IndexedIconName
        ExactWindowName
        ExactIconName

  - New Resize argument to direction:  "automatic" which works out,
    automatically, which direction a window should be resized in
    based on which quadrant the pointer is in on the window -- this
    is best observed by the grid formed from "ResizeOutline", as the
    shape of this grid is used when calculating this.

    Furthermore, the existing option of "warptoborder" can be used in
    conjunction with "direction automatic" to warp the pointer to the
    appropriate window edge.

  - Couple of changes to the Move command:
        - The EWMH working area is now honoured by default.
        - To move a window ignoring the working area, the option
          "ewmhiwa" can be used, similar to how the Maximize
          command works.

  - New expansion placeholder $[pointer.screen] to return the screen number
    the pointer is on.

* Bug fixes:

    - A few minor ones; nothing user-visible per se.

-------------------------------------------------------------------
Changes in stable release 2.6.2 (06-Aug-2011)

* New features:

  - New MenuStyle "UniqueHotkeyActivatesImmediate" and its
    negation "!UniqueHotkeyActivatesImmediate" to retain the menu
    until the user presses one of the menu keybindings to enact
    that change in the case where there's only one item defined in
    the menu at the specified hotkey.

* Bug fixes:

  - Fix infinite loop crash when FvwmButtons quits/restarts with
    RootTransparent colorsets.
  - Fix setting window states across FVWM restarts where those
    windows do not set WM_COMMAND.
  - Windows are now correctly restored when coming out of
    fullscreen mode.

-------------------------------------------------------------------
Changes in stable release 2.6.1 (16-Apr-2011)

* New features:

  None.

* Bug fixes:

  - Make the version and check of Fribidi >= 0.19.2

modules/MvwmConsole/MvwmConsole.c
/* -*-c-*- */
/*
 * Mvwm command input interface.
 *
 * Copyright 1996, Toshi Isogai. No guarantees or warantees or anything
 * are provided. Use this program at your own risk. Permission to use
 * this program for any purpose is given,
 * as long as the copyright is kept intact.
 */

/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* FIXME: The signal handling of this module is largely broken */

#include "config.h"

#include "MvwmConsole.h"
#include "libs/fio.h"
#include "libs/mvwmsignal.h"

/* using ParseModuleArgs now */
static ModuleArgs* module;

int Fd[2];  /* pipe to mvwm */
int  Ns = -1; /* socket handles */
char Name[80]; /* name of this program in executable format */
char *S_name = NULL;  /* socket name */

void server(void);
RETSIGTYPE DeadPipe(int);
void ErrMsg(char *msg);
void SigHandler(int);

void clean_up(void)
{
	if (Ns != -1)
	{
		close(Ns);
		Ns = -1;
	}
	if (S_name != NULL)
	{
		unlink(S_name);
		S_name = NULL;
	}

	return;
}

/*
 *      signal handler
 */
RETSIGTYPE DeadPipe(int dummy)
{
	clean_up();
	exit(0);

	SIGNAL_RETURN;
}

RETSIGTYPE SigHandler(int dummy)
{
	clean_up();
	exit(0);

	SIGNAL_RETURN;
}

RETSIGTYPE ReapChildren(int sig)
{
	mvwmReapChildren(sig);

	SIGNAL_RETURN;
}

int main(int argc, char *argv[])
{
	char client[120];
	char **eargv;
	int i, j, k;
	char *xterm_pre[] = { "-title", Name, "-name", Name, NULL };
	char *xterm_post[] = { "-e", NULL, NULL };
	int  clpid;

	/* Why is this not just put in the initializer of xterm_a?
	 * Apparently, it is a non-standard extension to use a non-constant
	 * address (of client) in an initializer (of xterm_a).
	 */
	xterm_post[1] = client;

	module = ParseModuleArgs(argc, argv, 0); /* we don't use an alias */
	if (module == NULL)
	{
		fprintf(
			stderr, "MvwmConsole version "VERSION" should only be"
			" executed by mvwm!\n");
		exit(1);
	}
	strcpy(Name, module->name);
	/* construct client's name */
	strcpy(client, argv[0]);
	strcat(client, "C");
	eargv = xmalloc((argc+12)*sizeof(char *));
	/* copy arguments */
	eargv[0] = XTERM;
	j = 1;
	for (k = 0; xterm_pre[k] != NULL; j++, k++)
	{
		eargv[j] = xterm_pre[k];
	}
	for (i = 0; i< module->user_argc; i++)
	{
		if (!strcmp(module->user_argv[i], "-e"))
		{
			i++;
			break;
		}
		else if (!strcmp(module->user_argv[i], "-terminal"))
		{
			i++;
			if (i < module->user_argc)
				/* use alternative terminal emulator */
				eargv[0] = module->user_argv[i];
		}
		else
		{
			eargv[j++] = module->user_argv[i];
		}
	}
	for (k = 0; xterm_post[k] != NULL; j++, k++)
	{
		eargv[j] = xterm_post[k];
	}

	/* copy rest of -e args */
	for (; i<module->user_argc; i++, j++)
	{
		eargv[j-1] = module->user_argv[i];
	}
	eargv[j] = NULL;
	signal(SIGCHLD, ReapChildren);
	/* Dead pipes mean mvwm died */
	signal(SIGPIPE, DeadPipe);
	signal(SIGINT, SigHandler);
	signal(SIGQUIT, SigHandler);
	signal(SIGALRM, SigHandler);

	Fd[0] = module->to_mvwm;
	Fd[1] = module->from_mvwm;

	/* launch xterm with client */
	clpid = fork();
	if (clpid < 0)
	{
		ErrMsg("client forking");
	}
	else if (clpid  == 0)
	{
		execvp(*eargv, eargv);
		ErrMsg("exec");
	}
	/* tell mvwm we're running */
	SendFinishedStartupNotification(Fd);

	server();

	return 0;
}

/*
 * setup server and communicate with mvwm and the client
 */
void server(void)
{
	struct sockaddr_un sas, csas;
	int  len;
	socklen_t clen;     /* length of sockaddr */
	char buf[MAX_COMMAND_SIZE];      /*  command line buffer */
	char *tline;
	char ver[40];
	fd_set fdset;
	char *home;
	int s;
	int msglen;
	int rc;

	/* make a socket  */
	s = socket(AF_UNIX, SOCK_STREAM, 0);
	if (s < 0 )
	{
		ErrMsg("socket");
	}

	/* name the socket */
	home = getenv("MVWM_USERDIR");
	S_name = xmalloc(strlen(home) + sizeof(S_NAME) + 1);
	strcpy(S_name, home);
	strcat(S_name, S_NAME);

	sas.sun_family = AF_UNIX;
	strcpy(sas.sun_path, S_name);
	/* bind the above name to the socket: first, erase the old socket */
	unlink(S_name);
	len = sizeof(sas) - sizeof(sas.sun_path) + strlen(sas.sun_path);
	umask(0077);
	rc = bind(s, (struct sockaddr *)&sas, len);
	if (rc < 0)
	{
		ErrMsg("bind");
	}
	else
	{
		/* don't wait forever for connections */
		alarm(MVWMCONSOLE_CONNECTION_TO_SECS);
		/* listen to the socket */
		rc = listen(s, 5);
		alarm(0);
		if (rc < 0)
		{
			ErrMsg("listen");
		}
		/* accept connections */
		clen = sizeof(csas);
		Ns = accept(s, (struct sockaddr *)&csas, &clen);
		if (Ns < 0)
		{
			ErrMsg("accept");
		}
	}
	/* send config lines to Client */
	tline = NULL;
	mvwm_send(Ns, C_BEG, strlen(C_BEG), 0);
	GetConfigLine(Fd, &tline);
	while (tline != NULL)
	{
		if (strlen(tline) > 1)
		{
			mvwm_send(Ns, tline, strlen(tline), 0);
			mvwm_send(Ns, "\n", 1, 0);
		}
		GetConfigLine(Fd, &tline);
	}
	mvwm_send(Ns, C_END, strlen(C_END), 0);
	strcpy(ver, "*");
	strcpy(ver, module->name);
	strcat(ver, " version ");
	strcat(ver, VERSION VERSIONINFO);
	strcat(ver, "\n");
	mvwm_send(Ns, ver, strlen(ver), 0);

	while (1)
	{
		FD_ZERO(&fdset);
		FD_SET(Ns, &fdset);
		FD_SET(Fd[1], &fdset);

		mvwmSelect(FD_SETSIZE, SELECT_FD_SET_CAST &fdset, 0, 0, NULL);
		if (FD_ISSET(Fd[1], &fdset))
		{
			MvwmPacket* packet = ReadMvwmPacket(Fd[1]);

			if (packet == NULL)
			{
				clean_up();
				exit(0);
			}
			else
			{
				if (packet->type == M_PASS)
				{
					msglen = strlen(
						(char *)&(packet->body[3]));
					if (msglen > MAX_MESSAGE_SIZE-2)
					{
						msglen = MAX_MESSAGE_SIZE-2;
					}
					mvwm_send(
						Ns, (char *)&(packet->body[3]),
						msglen, 0);
				}
			}
		}
		if (FD_ISSET(Ns, &fdset))
		{
			int len;
			int rc;

			rc = mvwm_recv(Ns, buf, MAX_COMMAND_SIZE, 0);
			if (rc == 0)
			{
				/* client is terminated */
				clean_up();
				exit(0);
			}
			/* process the own unique commands */
			len = strlen(buf);
			if (len > 0 && buf[len - 1] == '\n')
			{
				buf[len - 1] = '\0';
			}
			/* send command */
			SendText(Fd, buf, 0);
		}
	}
}

/*
 * print error message on stderr and exit
 */
void ErrMsg(char *msg)
{
	fprintf(
		stderr, "%s server error in %s, errno %d: %s\n", Name, msg,
		errno, strerror(errno));
	clean_up();
	exit(1);
}
